Powershell Provisoner unexpected EOF

Provisoning any powershell script returns an EOF error.

Packer OS: CentOS7
Packer Version 1.4.0

the powershell script is the following, though any powershell script seems to be causing an EOF
```
$invocation = (Get-Variable MyInvocation).Value
$directorypath = Split-Path $invocation.MyCommand.Path

Get-WindowsFeature | Where-Object -FilterScript { $_.Name -like '*Windows-Defender*' -or $_.Name -like '*SMB1*' -or $_.Name -like '*NET-WCF*' } | Remove-WindowsFeature
```

```
==> amazon-ebs: Provisioning with Powershell...
==> amazon-ebs: Provisioning with powershell script: /local/packer/playbooks/windows/scripts/base_removefeatures.ps1
Build 'amazon-ebs' errored: unexpected EOF

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: unexpected EOF

==> Builds finished but no artifacts were created.
panic: runtime error: invalid memory address or nil pointer dereference
2019/05/24 14:52:22 packer: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xe91452]
2019/05/24 14:52:22 packer:
2019/05/24 14:52:22 packer: goroutine 1527 [running]:
2019/05/24 14:52:22 packer: github.com/hashicorp/packer/communicator/winrm.(*Communicator).Upload(0xc000627260, 0xc00058aca0, 0x17, 0x324d2e0, 0xc000       0e89c0, 0x0, 0x24e6720, 0x1)
2019/05/24 14:52:22 packer:     /Users/mmarsh/go/src/github.com/hashicorp/packer/communicator/winrm/communicator.go:133 +0x242
2019/05/24 14:52:22 packer: github.com/hashicorp/packer/packer/rpc.(*CommunicatorServer).Upload(0xc00023dce0, 0xc0006c6880, 0xc00063f790, 0x0, 0x0)
2019/05/24 14:52:22 packer:     /Users/mmarsh/go/src/github.com/hashicorp/packer/packer/rpc/communicator.go:297 +0x148
2019/05/24 14:52:22 packer: reflect.Value.call(0xc0004bcc60, 0xc00000f0c0, 0x13, 0x2d81824, 0x4, 0xc000737f18, 0x3, 0x3, 0xc00072c840, 0x5, ...)
2019/05/24 14:52:22 packer:     /usr/local/go/src/reflect/value.go:447 +0x461
2019/05/24 14:52:22 packer: reflect.Value.Call(0xc0004bcc60, 0xc00000f0c0, 0x13, 0xc0005c1718, 0x3, 0x3, 0xc00066ca08, 0xc000074000, 0xc0005c1730)
2019/05/24 14:52:22 packer:     /usr/local/go/src/reflect/value.go:308 +0xa4
2019/05/24 14:52:22 packer: net/rpc.(*service).call(0xc000592440, 0xc000594550, 0xc000657750, 0xc000657760, 0xc0004f9080, 0xc0002345a0, 0x2472a60, 0x       c0006c6880, 0x16, 0x24e6720, ...)
2019/05/24 14:52:22 packer:     /usr/local/go/src/net/rpc/server.go:384 +0x14e
2019/05/24 14:52:22 packer: created by net/rpc.(*Server).ServeCodec
2019/05/24 14:52:22 packer:     /usr/local/go/src/net/rpc/server.go:481 +0x440
2019/05/24 14:52:22 [INFO] (telemetry) ending amazon-ebs
2019/05/24 14:52:22 ui error: Build 'amazon-ebs' errored: unexpected EOF
2019/05/24 14:52:22 Builds completed. Waiting on interrupt barrier...
2019/05/24 14:52:22 machine readable: error-count []string{"1"}
2019/05/24 14:52:22 ui error:
==> Some builds didn't complete successfully and had errors:
2019/05/24 14:52:22 machine readable: amazon-ebs,error []string{"unexpected EOF"}
2019/05/24 14:52:22 ui error: --> amazon-ebs: unexpected EOF
2019/05/24 14:52:22 ui:
==> Builds finished but no artifacts were created.
2019/05/24 14:52:22 [INFO] (telemetry) Finalizing.
2019/05/24 14:52:22 /opt/packer/packer: plugin process exited
2019/05/24 14:52:22 packer: 2019/05/24 14:52:22 Retryable error: Error uploading script: unexpected EOF
2019/05/24 14:52:22 waiting for all plugin processes to complete...
2019/05/24 14:52:22 /opt/packer/packer: plugin process exited
```

provisioners chunk of file
```
  "provisioners": [
    {
      "type": "file",
      "source": "{{user `packer_path`}}/playbooks/windows/files/lgpo.exe",
      "destination": "c:\\Windows\\Temp\\deploy\\"
    },
    {
      "type": "file",
      "source": "{{user `packer_path`}}/playbooks/windows/policy/basepcidss3",
      "destination": "c:\\Windows\\Temp\\deploy\\"
    },
    {
      "type": "file",
      "source": "{{user `packer_path`}}/playbooks/windows/sysprep/base.xml",
      "destination": "C:\\ProgramData\\Amazon\\EC2-Windows\\Launch\\Sysprep\\Unattend.xml"
    },
    {
    "type": "file",
    "destination": "c:\\Windows\\Temp\\deploy\\",
    "source":"{{user `packer_path`}}/playbooks/windows/scripts/base_removefeatures.ps1"
    },
    {
    "type": "powershell",
    "remote_path": "c:\\Windows\\Temp\\deploy\\",
    "script":"{{user `packer_path`}}/playbooks/windows/scripts/base_removefeatures.ps1"
    }
```

--- a/communicator/winrm/communicator.go
+++ b/communicator/winrm/communicator.go
@@ -130,7 +130,11 @@ func (c *Communicator) Upload(path string, input io.Reader, fi *os.FileInfo) err
 	}
 	if strings.HasSuffix(path, `\`) {
 		// path is a directory
-		path += filepath.Base((*fi).Name())
+		if fi != nil {
+			path += filepath.Base((*fi).Name())
+		} else {
+			return fmt.Errorf("Was unable to infer file basename for upload.")
+		}
 	}
 	log.Printf("Uploading file to '%s'", path)
 	return wcp.Write(path, input)

--- a/communicator/winrm/communicator_test.go
+++ b/communicator/winrm/communicator_test.go
@@ -120,5 +120,25 @@ func TestUpload(t *testing.T) {
 	if downloadedPayload != PAYLOAD {
 		t.Fatalf("files are not equal: expected [%s] length: %v, got [%s] length %v", PAYLOAD, len(PAYLOAD), downloadedPayload, len(downloadedPayload))
 	}
+}
 
+func TestUpload_nilFileInfo(t *testing.T) {
+	wrm := newMockWinRMServer(t)
+	defer wrm.Close()
+
+	c, err := New(&Config{
+		Host:     wrm.Host,
+		Port:     wrm.Port,
+		Username: "user",
+		Password: "pass",
+		Timeout:  30 * time.Second,
+	})
+	if err != nil {
+		t.Fatalf("error creating communicator: %s", err)
+	}
+	file := "C:\\Temp\\"
+	err = c.Upload(file, strings.NewReader(PAYLOAD), nil)
+	if err == nil {
+		t.Fatalf("Should have errored because of nil fileinfo")
+	}
 }

--- a/provisioner/powershell/provisioner.go
+++ b/provisioner/powershell/provisioner.go
@@ -9,6 +9,7 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"path/filepath"
 	"sort"
 	"strings"
 	"time"
@@ -252,6 +253,14 @@ func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.C
 		ui.Say(fmt.Sprintf("Provisioning with powershell script: %s", path))
 
 		log.Printf("Opening %s for reading", path)
+		fi, err := os.Stat(path)
+		if err != nil {
+			return fmt.Errorf("Error stating powershell script: %s", err)
+		}
+		if strings.HasSuffix(p.config.RemotePath, `\`) {
+			// path is a directory
+			p.config.RemotePath += filepath.Base((fi).Name())
+		}
 		f, err := os.Open(path)
 		if err != nil {
 			return fmt.Errorf("Error opening powershell script: %s", err)
@@ -272,7 +281,7 @@ func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.C
 			if _, err := f.Seek(0, 0); err != nil {
 				return err
 			}
-			if err := comm.Upload(p.config.RemotePath, f, nil); err != nil {
+			if err := comm.Upload(p.config.RemotePath, f, &fi); err != nil {
 				return fmt.Errorf("Error uploading script: %s", err)
 			}
 

